File: libraries/ios/Sources/VoidhashCore/Storage/RecordStore.swift
import Foundation

/// What ``RecordStore/load()`` found.
public struct RecordStoreLoad: Sendable {
    /// The complete records, oldest first.
    public let lines: [String]
    /// A trailing partial record dropped because the process died mid-append.
    public let droppedPartialRecord: Bool
    /// The store exists but could not be read, so `lines` says nothing about what is on disk.
    ///
    /// A caller must not rewrite or clear the store off a failed load: doing so would delete
    /// records it never saw. Appending is still safe.
    public let readFailed: Bool

    public init(lines: [String], droppedPartialRecord: Bool = false, readFailed: Bool = false) {
        self.lines = lines
        self.droppedPartialRecord = droppedPartialRecord
        self.readFailed = readFailed
    }
}

/// Append-only persistence for a queue of newline-delimited JSON records.
///
/// `UserDefaults` rewrites its whole backing file on every commit, which is the wrong shape for a
/// queue receiving an event at a time. The analytics queue and the transaction outbox use this
/// instead: appends are cheap, and the file is rewritten only when records are acknowledged.
public protocol RecordStore: Sendable {
    /// Returns every complete record, oldest first, and whether a partial one was discarded.
    func load() async -> RecordStoreLoad
    /// Appends `lines` to the end of the store, durably. Returns whether the write succeeded.
    @discardableResult
    func append(_ lines: [String]) async -> Bool
    /// Atomically replaces the entire contents with `lines`.
    func replace(with lines: [String]) async
    /// Removes the store.
    func clear() async
}

/// ``RecordStore`` backed by a file in the app's Application Support directory.
///
/// All file IO runs on a private serial queue rather than on the cooperative pool: a queue write
/// is blocking, and blocking a cooperative thread starves unrelated SDK work.
public final class FileRecordStore: RecordStore, @unchecked Sendable {
    private let url: URL
    private let diagnostics: DiagnosticEmitter
    private let queue: DispatchQueue

    /// - Parameters:
    ///   - url: File the records are written to. The parent directory is created on demand.
    ///   - diagnostics: Receives `QUEUE_READ_FAILED` when a store cannot be read or was truncated.
    public init(url: URL, diagnostics: DiagnosticEmitter = DiagnosticEmitter(nil)) {
        self.url = url
        self.diagnostics = diagnostics
        queue = DispatchQueue(label: "com.voidhash.sdk.record-store.\(url.lastPathComponent)")
    }

    /// Store for `name` under `Application Support/voidhash/<namespace>/`.
    ///
    /// Returns `nil` when no such directory can be resolved, letting the caller fall back to a
    /// cache-adapter backed store rather than silently losing durability.
    public static func applicationSupport(
        namespace: String,
        name: String,
        diagnostics: DiagnosticEmitter = DiagnosticEmitter(nil)
    ) -> FileRecordStore? {
        guard
            let base = FileManager.default.urls(
                for: .applicationSupportDirectory, in: .userDomainMask
            ).first
        else {
            return nil
        }
        // The directory is created by the first write, off the caller's thread: this runs on the
        // client's init path, which may be the main thread.
        let directory = base.appendingPathComponent("voidhash/\(namespace)", isDirectory: true)
        return FileRecordStore(
            url: directory.appendingPathComponent("\(name).ndjson"), diagnostics: diagnostics)
    }

    public func load() async -> RecordStoreLoad {
        return await run {
            guard FileManager.default.fileExists(atPath: self.url.path) else {
                // No file is the ordinary empty case, not a failure.
                return RecordStoreLoad(lines: [])
            }
            let data: Data
            do {
                data = try Data(contentsOf: self.url)
            } catch {
                self.diagnostics.emit(
                    .cache, code: "QUEUE_READ_FAILED", operation: "queue.load",
                    message:
                        "Could not read \(self.url.lastPathComponent): \(error.localizedDescription)"
                )
                return RecordStoreLoad(lines: [], readFailed: true)
            }

            let text = String(decoding: data, as: UTF8.self)
            var lines = text.split(separator: "\n", omittingEmptySubsequences: true).map(
                String.init)

            // A file that does not end in a newline was cut short mid-append: the last record is
            // incomplete and would decode into garbage, so it is dropped and reported.
            let droppedPartial = !data.isEmpty && data.last != 0x0A
            if droppedPartial {
                if !lines.isEmpty {
                    lines.removeLast()
                }
                _ = self.repairPartialTail(data)
            }
            return RecordStoreLoad(lines: lines, droppedPartialRecord: droppedPartial)
        }
    }

    @discardableResult
    public func append(_ lines: [String]) async -> Bool {
        guard !lines.isEmpty else {
            return true
        }
        let payload = Data((lines.joined(separator: "\n") + "\n").utf8)
        return await run {
            guard self.repairPartialTailIfNeeded() else {
                return false
            }
            guard let handle = try? FileHandle(forWritingTo: self.url) else {
                return self.writeWholeFile(payload)
            }
            defer { try? handle.close() }
            do {
                try handle.seekToEnd()
                try handle.write(contentsOf: payload)
                // Without the sync the records live in the page cache; a kill loses exactly the
                // events the persistent queue exists to protect.
                try handle.synchronize()
                return true
            } catch {
                self.diagnostics.emit(
                    .cache, code: "QUEUE_WRITE_FAILED", operation: "queue.append",
                    message:
                        "Could not append to \(self.url.lastPathComponent): \(error.localizedDescription)"
                )
                return false
            }
        }
    }

    public func replace(with lines: [String]) async {
        await run {
            guard !lines.isEmpty else {
                try? FileManager.default.removeItem(at: self.url)
                return
            }
            // `.atomic` writes a temporary file and renames it, so a crash mid-compaction leaves
            // the previous complete queue rather than a half-written one.
            self.writeWholeFile(Data((lines.joined(separator: "\n") + "\n").utf8))
        }
    }

    public func clear() async {
        await run {
            try? FileManager.default.removeItem(at: self.url)
        }
    }

    @discardableResult
    private func writeWholeFile(_ payload: Data) -> Bool {
        do {
            try FileManager.default.createDirectory(
                at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
            try payload.write(to: url, options: .atomic)
            return true
        } catch {
            diagnostics.emit(
                .cache, code: "QUEUE_WRITE_FAILED", operation: "queue.write",
                message: "Could not write \(url.lastPathComponent): \(error.localizedDescription)")
            return false
        }
    }

    // Only the last byte is inspected: reading the whole queue back on every coalesced append
    // would make each write cost as much as a full compaction. The file is read in full only
    // when that byte proves the tail is torn.
    private func repairPartialTailIfNeeded() -> Bool {
        guard FileManager.default.fileExists(atPath: url.path) else {
            return true
        }
        do {
            let handle = try FileHandle(forReadingFrom: url)
            defer { try? handle.close() }
            let size = try handle.seekToEnd()
            guard size > 0 else {
                return true
            }
            try handle.seek(toOffset: size - 1)
            guard let last = try handle.read(upToCount: 1)?.first, last != 0x0A else {
                return true
            }
            try handle.seek(toOffset: 0)
            let data = try handle.readToEnd() ?? Data()
            return repairPartialTail(data)
        } catch {
            diagnostics.emit(
                .cache, code: "QUEUE_READ_FAILED", operation: "queue.repair",
                message: "Could not inspect \(url.lastPathComponent): \(error.localizedDescription)"
            )
            return false
        }
    }

    private func repairPartialTail(_ data: Data) -> Bool {
        guard !data.isEmpty, data.last != 0x0A else {
            return true
        }
        diagnostics.emit(
            .cache, code: "QUEUE_RECORD_TRUNCATED", operation: "queue.load",
            message: "Discarded a partially written record at the end of \(url.lastPathComponent)"
        )
        let complete = data.lastIndex(of: 0x0A).map { Data(data.prefix(through: $0)) } ?? Data()
        return writeWholeFile(complete)
    }

    private func run<Value: Sendable>(_ work: @escaping @Sendable () -> Value) async -> Value {
        return await withCheckedContinuation { continuation in
            queue.async {
                continuation.resume(returning: work())
            }
        }
    }
}

/// ``RecordStore`` backed by a ``CacheAdapter`` entry holding the whole NDJSON blob.
///
/// Used where no writable file system directory is available and by tests that want a queue that
/// survives a client instance without touching disk.
public actor CacheAdapterRecordStore: RecordStore {
    private let adapter: any CacheAdapter
    private let key: String
    // The adapter read suspends the actor, so two appends racing through `load` would each
    // extend the same snapshot and the second write would drop the first's lines. The lines are
    // read once and every mutation happens on the in-memory copy before its write is issued.
    private var lines: [String]?
    private var loadTask: Task<[String], Never>?

    public init(adapter: any CacheAdapter, key: String) {
        self.adapter = adapter
        self.key = key
    }

    public func load() async -> RecordStoreLoad {
        return RecordStoreLoad(lines: await currentLines())
    }

    @discardableResult
    public func append(_ newLines: [String]) async -> Bool {
        guard !newLines.isEmpty else {
            return true
        }
        let merged = await currentLines() + newLines
        lines = merged
        await adapter.set(key, value: merged.joined(separator: "\n"))
        return true
    }

    public func replace(with newLines: [String]) async {
        lines = newLines
        guard !newLines.isEmpty else {
            await adapter.delete(key)
            return
        }
        await adapter.set(key, value: newLines.joined(separator: "\n"))
    }

    public func clear() async {
        lines = []
        await adapter.delete(key)
    }

    private func currentLines() async -> [String] {
        if let lines {
            return lines
        }
        let task =
            loadTask
            ?? Task { [adapter, key] in
                guard let raw = await adapter.get(key) else {
                    return [String]()
                }
                return raw.split(separator: "\n", omittingEmptySubsequences: true).map(String.init)
            }
        loadTask = task
        let loaded = await task.value
        // Whichever awaiter resumes first seeds the in-memory copy; the others find it set.
        if lines == nil {
            lines = loaded
        }
        loadTask = nil
        return lines ?? loaded
    }
}

/// ``RecordStore`` that keeps everything in memory.
public actor InMemoryRecordStore: RecordStore {
    private var lines: [String] = []

    public init() {}

    public func load() async -> RecordStoreLoad {
        return RecordStoreLoad(lines: lines)
    }

    @discardableResult
    public func append(_ newLines: [String]) async -> Bool {
        lines.append(contentsOf: newLines)
        return true
    }

    public func replace(with newLines: [String]) async {
        lines = newLines
    }

    public func clear() async {
        lines = []
    }
}

File: libraries/android/sdk/src/main/java/com/voidhash/sdk/storage/RecordStore.kt
package com.voidhash.sdk.storage

import com.voidhash.sdk.diagnostics.DiagnosticEmitter
import com.voidhash.sdk.diagnostics.VoidhashDiagnosticKind
import java.io.File
import java.io.FileOutputStream
import java.io.RandomAccessFile

/** Result of loading a record store, distinguishing an empty store from a failed read. */
data class RecordStoreLoad(
    val records: List<String>,
    val readFailed: Boolean = false,
)

/**
 * An append-only, line-delimited record store.
 *
 * `SharedPreferences` rewrites its whole file on every commit, which makes it the wrong
 * home for a thousand-entry queue that grows one event at a time. Queues therefore live in
 * their own NDJSON file: appends cost one write of the new lines, and the file is only
 * rewritten when the queue is compacted after a flush.
 *
 * Implementations are safe to call from several threads.
 */
interface RecordStore {
    /** Reads every stored record, oldest first. Unreadable lines are skipped. */
    fun readAll(): List<String>

    /** Loads every record while preserving whether storage could be read at all. */
    fun load(): RecordStoreLoad = RecordStoreLoad(readAll())

    /** Appends [records] to the end of the store, returning whether they reached storage. */
    fun append(records: List<String>): Boolean

    /** Rewrites the store to exactly [records], returning whether the rewrite succeeded. */
    fun replaceAll(records: List<String>): Boolean
}

/** In-memory [RecordStore] for tests and for clients configured with `enabled = false`. */
class InMemoryRecordStore(initial: List<String> = emptyList()) : RecordStore {
    private val records = ArrayList(initial)

    override fun readAll(): List<String> = synchronized(records) { records.toList() }

    override fun load(): RecordStoreLoad = RecordStoreLoad(readAll())

    override fun append(records: List<String>): Boolean {
        if (records.isEmpty()) return true
        synchronized(this.records) { this.records.addAll(records) }
        return true
    }

    override fun replaceAll(records: List<String>): Boolean {
        synchronized(this.records) {
            this.records.clear()
            this.records.addAll(records)
        }
        return true
    }
}

/**
 * [RecordStore] persisting one record per line in [file].
 *
 * A record containing a newline would break the format, so newlines are escaped on write
 * and restored on read. IO failures are reported through [onError] and never thrown: a
 * queue that cannot reach the disk still works in memory for the rest of the process.
 */
class FileRecordStore(
    private val file: File,
    private val diagnostics: DiagnosticEmitter = DiagnosticEmitter(),
    private val onError: (String, Throwable) -> Unit = { _, _ -> },
) : RecordStore {
    private fun report(context: String, error: Throwable) {
        diagnostics.emit(
            VoidhashDiagnosticKind.CACHE,
            code = "QUEUE_IO_FAILED",
            operation = "storage.$context",
            retryable = true,
            message = "Failed to $context: ${error.message}",
        )
        onError(context, error)
    }

    private val lock = Any()

    override fun readAll(): List<String> = load().records

    override fun load(): RecordStoreLoad = synchronized(lock) {
        if (!file.exists()) return RecordStoreLoad(emptyList())
        try {
            repairPartialTail()
            RecordStoreLoad(file.readLines().filter { it.isNotBlank() }.map(::decode))
        } catch (error: Throwable) {
            report("read ${file.name}", error)
            RecordStoreLoad(emptyList(), readFailed = true)
        }
    }

    override fun append(records: List<String>): Boolean {
        if (records.isEmpty()) return true
        return synchronized(lock) {
            try {
                file.parentFile?.mkdirs()
                repairPartialTail()
                file.appendText(records.joinToString(separator = "", transform = ::encode))
                true
            } catch (error: Throwable) {
                report("append to ${file.name}", error)
                false
            }
        }
    }

    override fun replaceAll(records: List<String>): Boolean {
        return synchronized(lock) {
            try {
                file.parentFile?.mkdirs()
                if (records.isEmpty()) {
                    file.writeText("")
                } else {
                    // Written through a temporary file so a crash mid-rewrite cannot leave a
                    // half-compacted queue behind.
                    val temporary = File(file.parentFile, "${file.name}.tmp")
                    FileOutputStream(temporary).use { output ->
                        output.write(
                            records.joinToString(separator = "", transform = ::encode)
                                .toByteArray(Charsets.UTF_8),
                        )
                        // Without this the rename can land before the bytes do, and a power
                        // loss leaves a correctly named, empty queue behind.
                        output.fd.sync()
                    }
                    if (!temporary.renameTo(file)) {
                        file.writeText(temporary.readText())
                        temporary.delete()
                    }
                }
                true
            } catch (error: Throwable) {
                report("rewrite ${file.name}", error)
                false
            }
        }
    }

    private fun encode(record: String): String = record
        .replace("\\", "\\\\")
        .replace("\n", "\\n")
        // `readLines` splits on a bare carriage return too, so it has to be escaped as well.
        .replace("\r", "\\r") + "\n"

    /** Removes bytes left by an append that died before writing its final newline. */
    private fun repairPartialTail() {
        if (!file.exists() || file.length() == 0L) return
        RandomAccessFile(file, "rw").use { handle ->
            handle.seek(handle.length() - 1)
            if (handle.read() == '\n'.code) return

            var offset = handle.length() - 1
            while (offset >= 0) {
                handle.seek(offset)
                if (handle.read() == '\n'.code) break
                offset -= 1
            }
            handle.setLength(offset + 1)
            handle.fd.sync()
        }
    }

    /**
     * Reverses [encode] in one pass.
     *
     * Two sequential replacements would be wrong: unescaping `\n` first turns the second
     * half of an escaped backslash-then-n into a real newline.
     */
    private fun decode(line: String): String {
        val decoded = StringBuilder(line.length)
        var index = 0
        while (index < line.length) {
            val character = line[index]
            if (character == '\\' && index + 1 < line.length) {
                when (line[index + 1]) {
                    'n' -> { decoded.append('\n'); index += 2; continue }
                    'r' -> { decoded.append('\r'); index += 2; continue }
                    '\\' -> { decoded.append('\\'); index += 2; continue }
                }
            }
            decoded.append(character)
            index += 1
        }
        return decoded.toString()
    }
}
